custom_status_bar - Allow for new, reasonably negative timers

leftos opened the pull request:
As discussed in https://github.com/cgtrout/MSFS-PyScriptManager/issues/49#issuecomment-2605917497, if the user has opted to allow for negative timers, we should support more than just allowing a positive timer to trickle into the negative, but rather also for timers to be created already expired / in the negative, as long as they're within a reasonable threshold (which is user-customizable via the function argument).

leftos committed:
custom_status_bar - If simbrief settings allow for a negative remaining timer, then accept a "future time" that's in the past

leftos committed:
custom_status_bar: Add negative timer threshold to compute_countdown_timer, so that new, reasonably delayed timers are allowed to start in the negative

cgtrout commented:
I've been testing this change with my test script and noticing incorrect behavior related to simulated UTC date crossovers with this change.  

However, when I think back to when I was building the 'midnight crossover' logic in compute_countdown_timer, I was building this functionality before I was aware of the MSFS 2024 bug.  Therefore, it is entirely possible the midnight crossover logic is completely unnecessary.

So this:
``` python

# Handle midnight rollover logic
if target_time_today < current_sim_time:
    print_debug("Target time is earlier than current simulator time.")
    if not is_negative and (last_countdown_time is None or last_countdown_time > 5):
        print_debug("Midnight rollover detected. Adjusting target time to next day.")
        target_time_today += timedelta(days=1)
```
may be completely unnecessary?

With this removed, the time calculation with this would need to be changed to:    
``` python
# Calculate remaining time
remaining_time = target_time - current_sim_time
```
I think as long as the date times are 'trusted' to be correct this should simplify the logic here immensely and calculate the correct value?

With this we have two options - I can apply this change and you can rebase to my change or you can make these changes directly into your branch?  I'm fine with either.

If you decide to make the change yourself, you can leave the function otherwise as it is - I can simplify the function signature and return values after.

leftos commented:
Doesn't that require the script not to ever assume the date, and always use datetime for both current sim time and target time? Current sim time now we have figured out how to get the date for, but when you ask the user for the "future time" in the dialog, will you have to also ask for the date? Or will you use some threshold to figure out if, for example, they mean 0425 UTC today, or 0425 UTC tomorrow?

That's the problem with trying to offer ease of use by allowing the date to be implied. Inference requires you to draw the line as to what is "too far in the past", at which point you can relatively safely assume it's meant to be a future time.

I get that that was an easy thing to answer before I asked you to allow for negative timers, but now...?

cgtrout commented:
This is the current function that adds a date to the typed string (part of CountdownTimerDialog):
``` python
def calculate_future_time(self, time_text):
        """
        Convert HHMM time input into a datetime object.
        Adjust for the next day if the entered time is earlier than the current simulator time.
        """
        hours, minutes = int(time_text[:2]), int(time_text[2:])
        current_sim_time = get_simulator_datetime()
        future_time = datetime(
            year=current_sim_time.year,
            month=current_sim_time.month,
            day=current_sim_time.day,
            hour=hours,
            minute=minutes,
            tzinfo=timezone.utc
        )

        if future_time < current_sim_time:
            future_time += timedelta(days=1)

        return future_time
```

I've been testing this with a variety of simulated simulator and realworld date-times and it seems to work as expected in all scenarios I've tested.

This will not allow a negative time (to be calculated), but I think for manual entry the user unlikely needs a negative time to be calculated?

Are you able to spot any potential edge cases on this I've missed maybe? - thanks

leftos committed:
Merge remote-tracking branch 'upstream/enhance/custom_bar-dynamic-simbrief' into custom_status_bar-negative_timer

leftos commented:
I was also talking about this in `compute_countdown_timer`:
```
    # Replace date for same-day calculation
    target_time_today = target_time.replace(
        year=current_sim_time.year, month=current_sim_time.month, day=current_sim_time.day
    )
    ```

leftos committed:
Simplify and fix compute_countdown_timer

- Trust inputs to the function to be complete datetimes, not just times that need to have their date adjusted
- Don't do any finagling with the date if it's determined to be in the past

This should hopefully allow for both negative timers as well as smooth midnight UTC rollover transitions

leftos commented:
Pull request updated based on the discussion. I was able to see my OFP from yesterday get pulled and successfully determine and show that I'm 16 hours past my gate out time (hah!), and hopefully by removing all the date finagling we'll have smooth midnight UTC rollover transitions as well.

leftos commented:
I just watched a midnight UTC rollover happen, and the remaining timer continued looking correct across it, fwiw.

leftos committed:
custom_status_bar: Don't display a huge remaining time if sim is not connected

leftos commented:
Noticed that when I reloaded the script with the sim not running, the current_sim_time fallback is UNIX_EPOCH, so I was getting a huge remaining value, hence the fix above.

cgtrout commented:
These changes all look solid - thanks for the help on this.  

When I wanted to collaborate on the functions in this manner I wasn't anticipating you needing to change this much - the assistance is greatly appreciated!

I will now bring this into the branch.

leftos commented:
No worries, happy to help.
